File: web/src/lib/decisions.ts
/** Owner decisions for edge cases — DB-backed so it survives serverless.
 *
 * When the agent hits a slot outside your green zone, the webhook (one
 * invocation) polls the DB while you tap a choice in the push notification. Your
 * tap (the /decide route — a SEPARATE invocation) writes the choice to the
 * call's `decision` column; the webhook's poll picks it up and books it.
 *
 * The old in-memory Map only worked in a single Node process; this works across
 * instances (Vercel/serverless), at the cost of ~1 DB read/sec while holding.
 */
import { prisma } from "./db";

export type Decision = { choice: string };

const POLL_MS = 1000;

/** True while the call is still awaiting the owner's tap (no decision yet). */
export async function isPending(key: string): Promise<boolean> {
  const row = await prisma.call.findUnique({
    where: { id: key },
    select: { status: true, decision: true },
  });
  return !!row && row.status === "awaiting_confirmation" && !row.decision;
}

/** Record the owner's tap. Returns false if the window already closed. */
export async function resolveDecision(key: string, decision: Decision): Promise<boolean> {
  if (!(await isPending(key))) return false;
  await prisma.call.update({
    where: { id: key },
    data: { decision: JSON.stringify(decision) },
  });
  return true;
}

/** Webhook side: poll until the owner taps or we time out. Consumes the
 *  decision (clears the column) so it can't be read twice. */
export async function waitForDecision(
  key: string,
  timeoutMs: number,
): Promise<Decision | null> {
  const deadline = Date.now() + timeoutMs;
  for (;;) {
    const row = await prisma.call.findUnique({
      where: { id: key },
      select: { decision: true },
    });
    if (row?.decision) {
      await prisma.call.update({ where: { id: key }, data: { decision: null } });
      try {
        return JSON.parse(row.decision) as Decision;
      } catch {
        return null;
      }
    }
    if (Date.now() >= deadline) return null;
    await new Promise((r) => setTimeout(r, POLL_MS));
  }
}

File: python/app/decisions.py
"""In-flight owner decisions for edge cases.

When the agent hits a slot outside your green zone, its tool call blocks here
waiting for you to tap a choice in the push notification. Your tap (the /decide
endpoint) resolves the awaiting future, and the agent — still on the line —
books the slot you picked. All within one call.

Futures live on the running asyncio loop, shared across the webhook request and
the /decide request. Single-process only; for multi-worker you'd back this with
Redis pub/sub.
"""
from __future__ import annotations

import asyncio
from typing import Dict, Optional

# call_record_id -> Future resolving to a decision dict
_PENDING: Dict[str, asyncio.Future] = {}


def open_decision(key: str) -> asyncio.Future:
    loop = asyncio.get_event_loop()
    fut: asyncio.Future = loop.create_future()
    _PENDING[key] = fut
    return fut


async def wait_for_decision(key: str, timeout: float) -> Optional[dict]:
    fut = _PENDING.get(key) or open_decision(key)
    try:
        return await asyncio.wait_for(fut, timeout=timeout)
    except asyncio.TimeoutError:
        return None
    finally:
        _PENDING.pop(key, None)


def resolve(key: str, decision: dict) -> bool:
    """Called by /decide when the owner taps. Returns True if someone was waiting."""
    fut = _PENDING.get(key)
    if fut and not fut.done():
        fut.set_result(decision)
        return True
    return False


def is_pending(key: str) -> bool:
    fut = _PENDING.get(key)
    return bool(fut and not fut.done())
